src/Channels.Networking.Sockets/SocketConnection.cs

```diff
@@ -398,28 +429,28 @@ private enum BufferStyle
         }
 
 
-        private const int SmallBufferBytes = 8;
-        private static byte[] LeaseSmallBuffer()
+        private static ArraySegment<byte> LeaseSmallBuffer()
         {
-            return Interlocked.Exchange(ref _recycledSmallBuffer, null) ?? new byte[SmallBufferBytes];
+            ArraySegment<byte> result;
+            if (!_smallBuffers.TryTake(out result))
+            {
+                // use a throw-away buffer as a fallback
+                result = new ArraySegment<byte>(new byte[_smallBuffers.BytesPerItem]);
+            }
+            return result;
         }
-        private void RecycleSmallBuffer(ref byte[] buffer)
+        private void RecycleSmallBuffer(ref ArraySegment<byte> buffer)
         {
-            if (buffer != null)
+            if (buffer.Array != null)
             {
-                // this is used as a sentinel in the unknown-strategy code;
-                // don't recycle it
-                if ((object)buffer != (object)_zeroLengthBuffer)
-                {
-                    Interlocked.Exchange(ref _recycledSmallBuffer, buffer);
-                }
-                buffer = null;
+                _smallBuffers?.Recycle(buffer);
```

Comment:
Why is there a null check here? Is it because these calls appear in finally blocks?